File: katas/4 kyu/range-extraction/solution.php
<!-- range-extraction -->
<?php
use PHPUnit\Framework\TestCase;
/* ------------------------------- MY SOLUTION ------------------------------ */
function solution(array $list): string
{
  $str = '';
  for($i = 0; $i < count($list) - 1;){
    if ($list[$i] + 1 < $list[$i + 1]){
      $str .= $list[$i] . ',';
      $i++;
    }else{
      $str .= $list[$i];
      $val = $list[$i];
      while ($i < count($list) - 1 && $list[$i] + 1 == $list[$i + 1]){
        $i++;
      }
      $str .= $list[$i] - $val == 1 ? ',' : '-';
    }
  }
  return $str . $list[count($list) - 1];
}
/* ----------------------------- CLEVER SOLUTION ---------------------------- */
/* ---------------------------------- TEST ---------------------------------- */
class ExampleTest extends TestCase
{
    /**
     * @test
     */
    public function example(): void
    {
        self::assertSame(
            '-6,-3-1,3-5,7-11,14,15,17-20',
            solution([-6, -3, -2, -1, 0, 1, 3, 4, 5, 7, 8, 9, 10, 11, 14, 15, 17, 18, 19, 20])
        );
    }
}

File: katas/4 kyu/range-extraction/solution.py
# range-extraction
import codewars_test as test
# -------------------------------- MY SOLUTION ------------------------------- #
from itertools import groupby
def group_consecutives(lst):
    for _, g in groupby(enumerate(lst), lambda i_x : i_x[0] - i_x[1]):
        r = [x for _, x in g]
        if len(r) > 2:
            yield f"{r[0]}-{r[-1]}"
        else:
            yield from map(str, r)

def solution(lst):
    return ','.join(group_consecutives(lst))
# ----------------------------------- TEST ----------------------------------- #
test.it("Simple Tests")
test.assert_equals(solution([-6,-3,-2,-1,0,1,3,4,5,7,8,9,10,11,14,15,17,18,19,20]), '-6,-3-1,3-5,7-11,14,15,17-20')
test.assert_equals(solution([-3,-2,-1,2,10,15,16,18,19,20]), '-3--1,2,10,15,16,18-20')